### PR TITLE
USearch Integration

### PR DESCRIPTION
**Pull request by davvard**
This PR implements the USearch integration. It's already feature-compatible with Annoy-backed vector-search, but can be easily extended to support more metrics, and allow on-the-fly quantization.

### CODE REVIEW

**Comment by rschu1ze** [comment]
Copypasting some notes from elsewhere:

"There are a few things to polish:
- Include paths of submodules in cmake
- While serializing USearch we can reserve needed buffer capacity ahead of time, our current approach is incorrect, suggestions welcomed
- Potential format mismatch between your and my versions of clang-format"

**Comment by rschu1ze** [comment]
@davvard Will look at this PR now but may I ask you first to
1. close/archive your existing fork of ClickHouse which forks [my fork](https://github.com/davvard/ClickHouse/tree/hnsw-index) of the upstream ClickHouse repository
2. make a new fork of ClickHouse based on the upstream repository ([here](https://github.com/ClickHouse/ClickHouse)).
3. squash all existing commits in this PR and push them to the new fork (previous step)
4. open a pull request which merges all changes against the upstream ClickHouse repository
5. I am happy to check this new PR then

The main reason I am asking (besides a simpler code flow) is that my fork of ClickHouse (current PR) has no CI tests configured, we would only know about issues with tests when I'd make another PR from my fork to to upstream.

**Comment by rschu1ze** [comment]
It would be nice if you could also update the docs in docs/en/engines/table-engines/mergetree-family/annindexes.md.

File: src/Storages/MergeTree/MergeTreeIndices.cpp [hunk 1]
```diff
@@ -1,9 +1,9 @@
-#include <Storages/MergeTree/MergeTreeIndices.h>
-#include <Parsers/parseQuery.h>
-#include <Parsers/ParserCreateQuery.h>
-#include <IO/WriteHelpers.h>
-#include <IO/ReadHelpers.h>
 #include <numeric>
+#include <IO/ReadHelpers.h>
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndices.cpp:2]: It's good practice to #include the corresponding header file (`MergeTreeIndices.h`) first.

EDIT: I guess that was clang-format's fault. Please turn it off (*) or make it use the .clang-format file in the top-level directory of the ClickHouse repository.

(*) ClickHouse CI is quite relaxed about code formatting. It only complains about rough misformatting (trailing spaces, braces not being on a new line) but it does not run clang-format.

File: src/Storages/MergeTree/MergeTreeIndexHnsw.h [hunk 2]
```diff
@@ -2,109 +2,102 @@
...
+    using Base = unum::usearch::index_dense_t;
+
+public:
+    explicit USearchIndexWithSerialization(size_t dimensions);
+    void serialize(WriteBuffer & ostr) const;
+    void deserialize(ReadBuffer & istr);
+    size_t getDimensions() const;
+};
+
+template <unum::usearch::metric_kind_t metric_ak>
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexHnsw.h:24]: Minor: What does `ak` stand for? Would just `metric` work as well?

File: src/Storages/MergeTree/MergeTreeIndexHnsw.cpp [hunk 3]
```diff
@@ -1,391 +1,354 @@
...
+#    include <DataTypes/DataTypeTuple.h>
+#    include <IO/ReadHelpers.h>
+#    include <IO/WriteHelpers.h>
+#    include <Interpreters/Context.h>
+#    include <Interpreters/castColumn.h>
+#    include <Storages/MergeTree/MergeTreeIndexHnsw.h>
+#    include <Common/typeid_cast.h>
+
+
+using namespace unum::usearch;
```
Comment (rschu1ze) [inline src/Storages/MergeTree/MergeTreeIndexHnsw.cpp:15]: Maybe a matter of personal preference, but I generally prefer not using `using namespace` as it hides some details. (and besides `using namespace`, I still find a lot of `unum::usearch::...` in this file)

**Review by rschu1ze — COMMENTED** [review]
Looks good upon a first look. That the implementation on ClickHouse side is very similar to Annoy certainly helps :-)

File: src/Storages/MergeTree/MergeTreeIndexHnsw.h (hunk 2 above)
Comment (davvard) [reply src/Storages/MergeTree/MergeTreeIndexHnsw.h:24]: OK I will write `metric` in the new PR

File: src/Storages/MergeTree/MergeTreeIndexHnsw.cpp (hunk 3 above)
Comment (davvard) [reply src/Storages/MergeTree/MergeTreeIndexHnsw.cpp:15]: Yes, I prefer not to use `using namespace` too. It appeared by accident

**Pull request closed without merging**